### PR TITLE
save-modal save button

### LINKED ISSUE #11963: Saving button can be pressed twice, generating an error
When trying to save a dashboard, a dialog box opens
image
If you click Save, nothing tells you that you clicked. If the operation last some seconds and you click twice (because you are not sure if you clicked or not). Then the operation is sent twice and an error arises:

image

I assume that this is because the second operation detects a conflict with the first one.

If the dashboard is waiting for metrics, the page waits to gather them all before saving, so the time to complete the save can last several seconds. 

I think that an easy solution could be to disable the save button just after clicking and changing the legend to "Saving...".

### PR DESCRIPTION
**Pull request by Ijin08**
added two spans to button to switch between, 
added class to succes-btn, 
created function that switch span and add class

fixes #11963

### CODE REVIEW

File: public/app/features/dashboard/save_modal.ts [hunk 1]
```diff
@@ -50,25 +53,35 @@ export class SaveDashboardModalCtrl {
...
+
     var dashboard = this.dashboardSrv.getCurrent();
     var saveModel = dashboard.getSaveModelClone();
     var options = { message: this.message };
 
     return this.dashboardSrv.save(saveModel, options).then(this.dismiss);
   }
+
+  setSaving() {
+    let saveButton = document.querySelector('#saveButton');
```
Comment (torkelo) [inline public/app/features/dashboard/save_modal.ts]: use ng-class in html template , example ng-class="{'btn--processing': ctrl.isSaving}" 

And the class needs to be more generic so it works for all types & themes, not sure if that is possible or just ticky. Also need a way to cancel isSaving state when errors happen. I wonder how we can do that without adding ugly try catch, and how we can make this more more reusable in other places.

**Review by torkelo — CHANGES_REQUESTED** [review]

**Review by marefr — CHANGES_REQUESTED** [review]
I can still click on *Saving...* and that triggers multiple calls for saving a dashboard.

Regarding Torkel's comment on error handling it looks like it's automatically solved since the save modal are closed on network error which is fine I guess.

File: public/app/features/dashboard/save_modal.ts [hunk 2]
```diff
@@ -50,8 +50,17 @@ const template = `
...
 
     <div class="gf-form-button-row text-center">
-      <button type="submit" class="btn btn-success" ng-disabled="ctrl.saveForm.$invalid">Save</button>
-      <a class="btn btn-link" ng-click="ctrl.dismiss();">Cancel</a>
+      <button
+        id="saveBtn"
+        type="submit"
+        class="btn btn-success"
+        ng-class="{'btn-success--processing': ctrl.isSaving}"
+        ng-disabled="ctrl.saveForm.$invalid"
```
Comment (marefr) [inline public/app/features/dashboard/save_modal.ts]: Did you try `ng-disabled="ctrl.saveForm.$invalid || ctrl.isSaving"`? Haven't tried it, but if that works I think you can remove the extra spans below and some logic in the controller.

File: public/app/features/dashboard/save_modal.ts [hunk 3]
```diff
@@ -120,6 +131,10 @@ export class SaveDashboardModalCtrl {
     var dashboard = this.dashboardSrv.getCurrent();
     var saveModel = dashboard.getSaveModelClone(options);
 
+    let myBtn = <HTMLInputElement>document.getElementById('saveBtn');
+    myBtn.disabled = true;
```
Comment (marefr) [inline public/app/features/dashboard/save_modal.ts]: This feels ugly - possible to use my suggested solution instead?

**Review by marefr — APPROVED** [review]
Great :+1:

**Pull request merged**